### PR TITLE
Add e2e tests to the Memcached Go sample providing an example and improved docs

### PR DESCRIPTION
**Pull request by camilamacedo86**
## Description
Init version of scaffold examples to let users know how they can write the e2e test for the Memcached Go

Note that to ensure that tests are added to the sample we are also adding them to the CI (`memcached-sample`)
However, these tests are lighter and we do not need to build the scorecard images and etc to run them

Also, its intention is not to validate all SDK features like scorecard as we do in the SDK e2e tests that are used to validate the changes on the project.

### CODE REVIEW

File: testdata/go/v3/memcached-operator/test/e2e/e2e_test.go
Comment (everettraven) [inline testdata/go/v3/memcached-operator/test/e2e/e2e_test.go:103]: Is this step necessary? This looks instead like it is pulling the Operand image to the host machine but not adding it to the KinD cluster

File: testdata/go/v3/memcached-operator/test/e2e/e2e_test.go
Comment (everettraven) [inline testdata/go/v3/memcached-operator/test/e2e/e2e_test.go:48]: Are these steps necessary? It doesn't seem like the test is doing anything regarding metrics. I am also not sure if anything that is being done in the test would require cert-manager.

**Review by everettraven — COMMENTED** [review]
I like the idea of providing an example of e2e tests.

I think this (in my mind) also begs the question: Should we be scaffolding a simple suite of tests (e2e & unit?) that covers some basic cases that a user can extend as they implement the functionality of their operator?

Regarding this PR, I have a couple questions:

File: testdata/go/v3/memcached-operator/test/e2e/e2e_test.go
Comment (camilamacedo86) [reply testdata/go/v3/memcached-operator/test/e2e/e2e_test.go:103]: we need to do both right? pull and load

Comment (camilamacedo86) [inline testdata/go/v3/memcached-operator/test/e2e/e2e_test.go:103]:
```suggestion
```

Comment (camilamacedo86) [reply testdata/go/v3/memcached-operator/test/e2e/e2e_test.go:103]: good catcher

File: Makefile
Comment (varshaprasad96) [inline Makefile:179]: Why create a separate target? Can't this be done along with test-e2e-go?
This is a part of testing scaffolded go memcached operator. Why not do it together?

File: testdata/go/v3/memcached-operator/test/e2e/e2e_test.go
Comment (varshaprasad96) [reply testdata/go/v3/memcached-operator/test/e2e/e2e_test.go:48]: A few questions:
1. For a beginner in operator-SDK do we need e2e tests with webhooks? IMO webhooks are an advanced topic. Can't we disable the webhooks and run. 
2. Also, `testdata/` is a the place where we intend that user looks for a sample operator. If we are adding these e2e tests, I think we should mention this somewhere in our docs too explaining what these e2e tests are intended for. Especially if we decide to have webhooks, a new user wouldn't be aware of these pre-requisites at all.

Comment (camilamacedo86) [reply testdata/go/v3/memcached-operator/test/e2e/e2e_test.go:48]: @varshaprasad96,
 
The samples are scaffolded with webhooks and with all options enable. So it is scaffold with cert-manager enable. We are not changing the sample here at all. We are only adding an e2e test as an example to show how the users can do that. 

(ihmo) The samples are examples are useful for:
- Help the users know how they can be built their solutions ( example )
- For the maintainer to check how the changes made in the tool affect the scaffolds, in this way also in the review and for we test the commands

The motivation here would be to link in the docs to show the author how they can test their solutions. I think we should have docs about that but it seems a gap for SDK and KB indeed. But we do things on top of the samples because they are our examples.

File: Makefile
Comment (camilamacedo86) [reply Makefile:179]: The target was created following the same structure as the others. See:

- test-e2e-go: it has the purpose to run the e2e test of the SDK project to provide projects that work and that the features provided by the CLI work well with ( scorecard, bundle, etc )
- test-e2e-sample-go: has the purpose of triggering the e2e tests scaffold only for the Memcached sample done with go/v3 

What was my motivation to create a separate target instead? 
- If the tests stop working/break is eaiser to identify what test has an issue that needs to be fixed as re-test locally only what is required to be tested ( faster ) 
- To run the test-e2e-go we need to build the scorecard images and etc which takes a long time when to run this one we just need to have a Kind up and running. They are lighter and much faster. 





I created a separate target so we are able to run only these checks if we need. IMO

**Review by everettraven — APPROVED** [review]
This looks good to me, but I think it would be best to have another set of eyes and approval before merging

File: hack/generate/samples/internal/go/v3/e2e_test_code.go
Comment (varshaprasad96) [inline hack/generate/samples/internal/go/v3/e2e_test_code.go:185]: Do we run `make uninstall` too? Envtest will clean up the cluster anyway by removing CRDs, but checking if we need that best practice in here too just so that users know.

File: website/content/en/docs/best-practices/common-recommendation.md
Comment (varshaprasad96) [inline website/content/en/docs/best-practices/common-recommendation.md:79]:
```suggestion
    - For Go-based operators, you can also use [envtest][envtest] to cover the controllers. For further information see [Testing with EnvTest][testing-with-envtest]. Also, see the `test` directory for the Memcached sample under the [testdata/go/v3/memcached-operator][sample] to know how can you build e2e tests.
```

Comment (varshaprasad96) [reply website/content/en/docs/best-practices/common-recommendation.md:79]: @camilamacedo86 Here is where I meant we could explain how we are setting up envtest. Similar to https://book.kubebuilder.io/reference/envtest.html#environment-variables. The KB doc does not mention about installing cert-manager and prometheus. We could add that here as an extension telling why we do so, and how we test webhooks and a running operator. This could be done in a follow up too.

**Review by varshaprasad96 — COMMENTED** [review]
Just a few follow ups. But agree that there are more questions coming in regarding testing of operators using envtest and this PR would definitely be useful!

File: hack/generate/samples/internal/go/v3/e2e_test_code.go
Comment (camilamacedo86) [reply hack/generate/samples/internal/go/v3/e2e_test_code.go:185]: @varshaprasad96 we need to create the NS first to be able to label it as restricted. So that we can check if the manifest will or not be applied and are working under restricted enforcement.

File: website/content/en/docs/best-practices/common-recommendation.md
Comment (camilamacedo86) [reply website/content/en/docs/best-practices/common-recommendation.md:79]: @varshaprasad96 that is agreat point. I think it needs to be in the kb doc right since are options provided by default. I tracked the need in:  https://github.com/kubernetes-sigs/kubebuilder/issues/2777.

But sure we can improve in follow-ups these points here as well.

File: hack/generate/samples/internal/go/v3/e2e_test_code.go
Comment (camilamacedo86) [reply hack/generate/samples/internal/go/v3/e2e_test_code.go:185]: @varshaprasad96 this test is done done with ENVTEST. Why? 
ENVTEST does not allow us to do this e2e. It is more for you to test just the controller and not the whole solution.

**Review by jmrodri — CHANGES_REQUESTED** [review]
minor change

File: testdata/go/v3/memcached-operator/test/e2e/e2e_test.go
Comment (camilamacedo86) [reply testdata/go/v3/memcached-operator/test/e2e/e2e_test.go:51]: I added comments that clarify.

**Review by jmrodri — APPROVED** [review]

**Review by varshaprasad96 — APPROVED** [review]
/lgtm

**Pull request merged**
